pkg/controller/master/node/delete_controller.go

```diff
@@ -0,0 +1,244 @@
+// - Delete the complete job `harvester-delete-node-<node-name>` and remove the finalizer.
+type deleteNodeHandler struct {
+	nodes                       ctlcorev1.NodeClient
+	nodeCache                   ctlcorev1.NodeCache
+	virtualMachineInstanceCache v1.VirtualMachineInstanceCache
+	jobs                        ctlbatchv1.JobClient
+	jobCache                    ctlbatchv1.JobCache
+	namespace                   string
+}
+
+// DeleteRegister registers the node controller
+func DeleteRegister(ctx context.Context, management *config.Management, options config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	jobs := management.BatchFactory.Batch().V1().Job()
+	vmis := management.VirtFactory.Kubevirt().V1().VirtualMachineInstance()
+	deleteNodeHandler := &deleteNodeHandler{
+		nodes:                       nodes,
+		nodeCache:                   nodes.Cache(),
+		jobs:                        jobs,
+		jobCache:                    jobs.Cache(),
+		virtualMachineInstanceCache: vmis.Cache(),
+		namespace:                   options.Namespace,
+	}
+
+	nodes.OnChange(ctx, deleteNodeControllerName, deleteNodeHandler.OnNodeChanged)
+	nodes.OnRemove(ctx, deleteNodeControllerName, deleteNodeHandler.OnNodeRemove)
+
+	jobs.OnChange(ctx, deleteNodeControllerName, deleteNodeHandler.OnJobChanged)
+
+	return nil
+}
+
+func (h *deleteNodeHandler) OnNodeChanged(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+
+	if !containsFinalizer(node, deleteNodeFinalizer) {
+		toUpdate := node.DeepCopy()
+		toUpdate.Finalizers = append(toUpdate.GetFinalizers(), deleteNodeFinalizer)
+		return h.nodes.Update(toUpdate)
+	}
+
+	return node, nil
+}
+
+func (h *deleteNodeHandler) OnNodeRemove(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil {
+		return node, nil
+	}
+
+	if node.DeletionTimestamp != nil {
+		if _, err := h.createDeleteNodeJob(node); err != nil {
+			return node, err
+		}
+	}
+
+	return node, nil
+}
+
+func (h *deleteNodeHandler) OnJobChanged(key string, job *batchv1.Job) (*batchv1.Job, error) {
+	if job == nil || job.DeletionTimestamp != nil {
+		return job, nil
+	}
+
+	nodeName, ok := job.Labels[HarvesterDeleteNodeLabelKey]
+	if !ok {
+		return job, nil
+	}
+
+	if ConditionJobComplete.IsTrue(job) {
+		node, err := h.nodeCache.Get(nodeName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				policy := metav1.DeletePropagationBackground
+				return job, h.jobs.Delete(job.Namespace, job.Name, &metav1.DeleteOptions{PropagationPolicy: &policy})
+			}
+			return job, err
+		}
+		if _, err := h.removeFinalizer(node); err != nil {
+			return job, err
+		}
+	}
+
+	return job, nil
+}
+
+func (h *deleteNodeHandler) createDeleteNodeJob(node *corev1.Node) (*batchv1.Job, error) {
+	job := createDeleteNodeJob(h.namespace, node)
+	return h.jobs.Create(job)
+}
+
+func (h *deleteNodeHandler) removeFinalizer(node *corev1.Node) (*corev1.Node, error) {
```

Comment:
We have the finalizer operation logic in the framework, can we use that?
https://github.com/rancher/wrangler/blob/master/pkg/generic/remove.go#L39-L56

In short, an onRemove handler adds the finalizer and removes it when it succeeds.

Reply:
So, for reusing the finalizer logic from the framework, the controller needs to emit an error for await the job removing RKE2 file. And then the job updates some label on the node in order to  re-run `OnRemove` handler.

It seems awkward to me if we emits an error only for the sake of reusing the implicit finalizer logic. Is it a conventional use case of extending finalizer `wrangler.cattle.io/xxxx`? My first impression is that it's a basic protection for any "real error" inside `OnRemove`, but not a fake error we are going to emit.

Reply:
Make sense. I'm good with the current approach